sycl/doc/extensions/proposed/sycl_ext_oneapi_user_defined_reductions.asciidoc

```diff
@@ -0,0 +1,188 @@
+  T reduce_over_group(GroupHelper g, V x, T init, BinaryOperation binary_op); // (2)
+}
+----
+
+**TODO**: consider adding `is_group_with_scratchpad_v` helper to describe
+preconditions better.
+
+1._Constraints_: `GroupHelper` is a `group_with_scratchpad`.
+
+_Mandates_: `binary_op(x, x)` must return a value of type `T`.
+
+_Preconditions_: The size of memory contained by `GroupHelper` object `g` must
+be at least `sizeof(T) * g.get_group().get_local_range().size()` bytes.
+`binary_op` must be an instance of a function object.
+
+_Returns_: The result of combining all the values of `x` specified by each
+work-item in the group using the operator `binary_op`, where the values are
+combined according to the generalized sum defined in standard C++.
+
+NOTE: If `T` is a fundamental type and `BinaryOperation` is a SYCL function
+object type, then memory attached to `GroupHelper` object `g` is not used and
+the call to this overload is equivalent to calling
+`reduce_over_group(g.get_group(), x, binary_op)`.
+
+2._Constraints_: `GroupHelper` is a `group_with_scratchpad`.
+
+_Mandates_: `binary_op(init, x)` and `binary_op(x, x)` must return a value of
+type `T`.
+
+_Preconditions_: The size of memory contained by `GroupHelper` object `g` must
+be at least `sizeof(T) * g.get_group().get_local_range().size()` bytes.
+`binary_op` must be an instance of a function object.
+
+_Returns_: The result of combining all the values of `x` specified by each
+work-item in the group and the initial value `init` using the operator
+`binary_op`, where the values are combined according to the generalized sum
+defined in standard C++.
+
+NOTE: If `T` and `V` are fundamental types and `BinaryOperation` is a SYCL
+function object type, then memory attached to `GroupHelper` object `g` is not
+used and the call to this overload is equivalent to calling
+`reduce_over_group(g.get_group(), x, init, binary_op)`.
+
+NOTE: Implementation of both overaloads may use less memory than passed
+to the function depending on the exact algorithm which is used for doing the
+reduction.
+
+== Example usage
+
+[source,c++]
+----
+template <typename T>
+struct UserDefinedSum {
+  T operator()(T a, T b) {
+    return a + b;
+  }
+};
+
+q.submit([&](sycl::handler& h) {
+  auto acc = sycl::accessor(buf, h);
+
+  constexpr size_t group_size = 256;
+
+  // Create enough local memory for the algorithm
+  size_t temp_memory_size = group_size * sizeof(T);
+  auto scratch = sycl::local_accessor<std::byte, 1>(temp_memory_size, h);
+
+  h.parallel_for(sycl::nd_range<1>{N, group_size}, [=](sycl::nd_item<1> it) {
+    // Create a handle that associates the group with an allocation it can use
+    auto handle = sycl::ext::oneapi::experimental::group_with_scratchpad(
+        it.get_group(), sycl::span(scratch, temp_memory_size));
+
+    // Pass the handle as the first argument to the group algorithm
+    T sum = sycl::reduce_over_group(handle, acc[it.get_global_id(0)], 0,
```

Comment:
```suggestion
    T sum = sycl::ext::oneapi::experimental::reduce_over_group(handle, acc[it.get_global_id(0)], 0,
```
?

Reply:
I think this is unnecessary, because the handle itself is already in the `experimental` namespace.

SYCL 2020 has some [guidelines for extensions](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_names_for_extensions_to_existing_classes_or_enumerations); while it doesn't specifically mention adding overloads to existing SYCL functions, I think the paragraph about adding constructors could apply here:

> In some cases, an extension does not have the freedom to choose a specific function name. For example, this could happen if the extension adds a new constructor overload for an existing SYCL class. In cases like this, the extension should ensure that one of the function parameters has a type that is defined in the extension’s namespace. For example, the Acme vendor could add a new constructor for `sycl::context` with the signature `context(ext::acme::frobber &)`.

A developer cannot invoke the new overload of `reduce_over_group` without first creating a handle, and there's no risk of collision.

Reply:
@Pennycook, do I understand it correctly, that I should update the extension then to put those functions _out_ of experimental namespace? 

I think what @dm-vodopyanov is trying to highlight here is that the example doesn't match with the specification above

Reply:
Oh, I'd missed that you'd defined the functions in the experimental namespace. Yes, my preference here would be to move the new overloads out into the `sycl` namespace.

This is simpler and allows generic programming: it allows developers to treat `group_with_scratchpad` as though it is just another group, and to pass that new group type to the existing `reduce_over_group` function.  Having two versions of `reduce_over_group` floating around in different namespaces might confuse people, especially if the version in `experimental` is _only_ compatible with `group_with_scratchpad` (i.e. it is incompatible with `group`, `sub_group`, etc).

I'd be interested to hear what @gmlueck thinks about this, though.

Reply:
The namespace does more than avoid collisions with other vendor's extensions.  It also makes it clear in the source code that the application is using something non-standard, which is something our customers said they wanted when we presented the extension mechanism at TAB.

Since this is an experimental extension, it's probably even more valuable to make it clear that this API is not only non-standard but also not suitable for production code.

I'd lean towards keeping these free functions in the `experimental` namespace unless there is a strong motivation for the generic programming use case.

Reply:
Just to make sure I understand... Why isn't the `group_with_scratchpad` being in `experimental` enough to make it clear that something non-standard is being used here?

I can see an argument that because we're extending the types supported by `reduce_over_group` rather than just adding an overload that accepts a `group_with_scratchpad`, it may still be valuable to convey that the whole function is experimental...  Would you feel differently if we were just adding support for a new group to the API?

I ask because we have a few other proposals adding group types (e.g. root-groups, the non-uniform sub-groups) and I'm wondering if we also need to redefine the algorithms in `experimental` for those groups.

Reply:
Yes, I would feel differently for an extension that adds a new group type that was expected to work with existing function templates.  In such a case, I think it would be fine for the existing function templates to use the new type.  In this case, we added new function overloads specifically for this extended type.  There doesn't seem to be any overriding reason to put those overloads directly in the `sycl::` namespace, so it seems appropriate to put them in the extension's namespace.

Reply:
Ok, that makes sense.  @AlexeySachkov: Please disregard my comments above and keep this overload in the experimental namespace.